### ISSUE TITLE
Improve performance by supporting multi-threaded compression

### ISSUE THREAD

**Issue by zylo117** [sergey-dryabzhinsky/python-zstd#48]
Thank you for your hard work.

I just wonder if python-zstd can be much faster by using multi-threaded compression?

I tried comprssing a 800MB+ data, it takes python-zstd 0.48s, faster than zstandard, which is 0.67s.

However, zstandard supports multi-threaded compression, and when I use 6 threads, it only takes 0.17s.

I don't really mean to compare or judge, but I personally believe your implement is more performant, and python-zstd can be more perfect if it supports multi-threaded compression. Thanks.

**Comment by sergey-dryabzhinsky** [comment]
If you use python-zstd with bundled libzstd then it should be builded with multi-thread support.

How do you process your test 800Mb of data? By many chunks or one brick?

**Comment by sergey-dryabzhinsky** [comment]
Ah, seems like libzstd works in single-thread mode by default.

I should rethink internal processes to use multi-thread abilities, use other API methods...

**Comment by zylo117** [comment]
> If you use python-zstd with bundled libzstd then it should be builded with multi-thread support.
> 
> How do you process your test 800Mb of data? By many chunks or one brick?

I just generated a numpy array that is big enough to test the speed, no chunking.

**Comment by zylo117** [comment]
Oh, I also monitored the processes/threads while compressing with python-zstd, there are 6 threads running in parallel (my cpu has 6 cores).

But only one thread is saturated, other 5 just sitting out.

**Comment by sergey-dryabzhinsky** [comment]
Could you try new code from `issue-48-use-multithread-api` branch?
Added new function `compress_mt`.
If it all ok - I'll make it default.

**Comment by zylo117** [comment]
> Could you try new code from `issue-48-use-multithread-api` branch?
> Added new function `compress_mt`.
> If it all ok - I'll make it default.

I have tried it, compress_mt didn't work, even if I set 'threads' to a specified number, like 10, there was still only one core working.

Did you succeed running it with multithread on your computer? Or was I compiling it wrong?

I followed the steps of your page and I did call function 'compress_mt' though, no exception was raised.

**Comment by sergey-dryabzhinsky** [comment]
```
>>> import zstd
>>> import uuid
>>> from time import time
>>> t=time(); c = zstd.compress(uuid.uuid4().bytes*1024*1024*100, 3); print(time()-t);
0.8209028244018555
>>> t=time(); c = zstd.compress_mt(uuid.uuid4().bytes*1024*1024*100, 3); print(time()-t);
0.8217425346374512
>>> t=time(); c = zstd.compress_mt(uuid.uuid4().bytes*1024*1024*100, 3, 0); print(time()-t);
0.8171048164367676
>>> t=time(); c = zstd.compress_mt(uuid.uuid4().bytes*1024*1024*100, 3, 1); print(time()-t);
0.81520676612854
>>> t=time(); c = zstd.compress_mt(uuid.uuid4().bytes*1024*1024*100, 3, 2); print(time()-t);
0.7785365581512451
>>> t=time(); c = zstd.compress_mt(uuid.uuid4().bytes*1024*1024*100, 3, 4); print(time()-t);
```
I think it's working. Maybe not so optimized.

**Comment by sergey-dryabzhinsky** [comment]
More accurate:
```
>>> import zstd
>>> import uuid
>>> from time import time
>>> d = uuid.uuid4().bytes*1024*1024*100
>>> t=time(); c = zstd.compress(d, 6); print(time()-t);
0.16142892837524414
>>> t=time(); c = zstd.compress_mt(d, 6); print(time()-t);
0.07590532302856445
>>> t=time(); c = zstd.compress_mt(d, 6, 4); print(time()-t);
0.07538676261901855
>>> t=time(); c = zstd.compress_mt(d, 6, 2); print(time()-t);
0.10687851905822754
>>> t=time(); c = zstd.compress_mt(d, 6, 1); print(time()-t);
0.16031503677368164
```

**Comment by sergey-dryabzhinsky** [comment]
Threads are quickly created and flushed.
If you want to see them use this code:
```
>>> for i in range(1000):
...     t=time(); c = zstd.compress_mt(d, 6); print(time()-t);
```

**Comment by sergey-dryabzhinsky** [comment]
Updated branch - now threads count is auto-tuned by CPU cores count.

**Comment by zylo117** [comment]
I updated your code and built again, and now it works, thanks.

When using compress_mt, it takes only 0.14s!!! Awesome! Thanks again!

**Comment by sergey-dryabzhinsky** [comment]
Okay, so new code will be default one.

**Comment by sergey-dryabzhinsky** [comment]
Done!
Released new version 1.4.4.1 with multi-threaded api enabled.
